Purchases: Add a billing period switch to PurchasePlanDetails

tyxla opened the pull request:
This PR introduces a Billing period switch to the plan details card for Jetpack plan purchases, as suggested in p6TEKc-17s-p2. It allows the user to switch from a monthly to yearly plan - after the user selects "yearly", they're redirected to the checkout page to complete the purchase. Downgrading from yearly to monthly is intentionally not supported for now.

This PR is intended to be landed together with D5920-code, which introduces discounts for the monthly-to-yearly upgrade case. Feel free to test with it to get a nice discount when upgrading from a monthly to an equivalent or higher tier yearly plan.

This PR implements part of #6499.

Preview (Jetpack monthly plan):
![](https://cldup.com/cRzrwjkHhr.png)

Preview (Jetpack yearly plan):
![](https://cldup.com/clWXZ-Pl3M.png)

To test:
* Checkout this branch or get it going on calypso.live
* Go to /me/purchases and click on a Jetpack monthly plan purchase.
* Verify you see a Billing period label with a select field in the plan details card.
* Select "yearly" and verify you're able to complete the purchase successfully. 
* Go to /me/purchases again and click on a Jetpack yearly plan purchase.
* Verify you see a Billing period label, but with a static "yearly" copy below.
* Go to /me/purchases again and select a WordPress.com plan.
* Verify plan details page is untouched and unaffected from this PR.

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,76 @@
+		if ( ! isMonthly( purchase.productSlug ) ) {
+			return (
+				<FormSettingExplanation>
+					{ translate( 'yearly' ) }
```
Hi! I've found a possible matching string that has already been translated **31** times:
`translate( 'Yearly' )` *ES Score: 18*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,76 @@
+	};
+
+	handleBillingPeriodChange = ( event ) => {
+		if ( event.target.value === 'monthly' ) {
```
Hi! I've found a possible matching string that has already been translated **32** times:
`translate( 'Monthly' )` *ES Score: 17*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-plan-purchases-billing-period-switch)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

rickybanister commented:
I know it will be redundant to the expiration date listed above, but I wonder if we shouldn't reiterate the renewal date for yearly plans in this location?

So instead of 'yearly' we could write `Billed yearly, renews August 26th 2017` or similar. 

Also I think we should capitalize 'monthly' and 'yearly' in the dropdown/read-only text. 

What do you think @richardmuscat

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,113 @@
+			} );
+		}
+
+		return translate( 'Billed yearly' );
```
Hi! I've found a possible matching string that has already been translated **22** times:
`translate( ', billed yearly' )` *ES Score: 20*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-plan-purchases-billing-period-switch)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,113 @@
+		}
+
+		if ( isExpired( purchase ) ) {
+			return translate( 'Billed yearly, expired %(timeSinceExpiry)s', {
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'Expired %(timeSinceExpiry)s', { context: 'timeSinceExpiry is of the form "[number] [time-period] ago" i.e. "3 days ago"'} )` *ES Score: 14*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

:information_source: This string is probably long enough that it doesn't need a **context**. Are you sure you don't want to use a translator comment instead?

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,113 @@
+		const { purchase, translate } = this.props;
+
+		if ( showCreditCardExpiringWarning( purchase ) ) {
+			return translate( 'Billed yearly, credit card expiring soon' );
```
Hi! I've found a possible matching string that has already been translated **23** times:
`translate( 'Credit card expiring soon' )` *ES Score: 8*

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

tyxla commented:
@rickybanister I've addressed your suggestions now:

>  I wonder if we shouldn't reiterate the renewal date for yearly plans in this location?

Done, here are some previews:

![](https://cldup.com/sTVkC6gv-p.png)
![](https://cldup.com/AbfF1u2_gR.png)

> Also I think we should capitalize 'monthly' and 'yearly' in the dropdown/read-only text.

Oh yes, I did this immediately after submitting the PR, thanks to @a8ci18n 😉 

![](https://cldup.com/z5R-vkD3M0.png)

This PR could use another code and design review now.

sirreal commented on client/lib/plans/constants.js:
```diff
@@ -1002,3 +1002,16 @@ export function getMonthlyPlanByYearly( plan ) {
 	}
 }
+
+export function getYearlyPlanByMonthly( plan ) {
```
It's a shame #14761 hasn't landed yet

There are a lot of helper functions that live in this `constants` file which doesn't seem correct. I think it would be better if we tried to keep these functions in a more appropriate place.

tyxla replied:
That's true. It's an improvement that is totally worth some time, but probably for another PR after we land #14761?

seear commented:
It seems slightly odd that changing the value of a dropdown whisks you off to another location (the cart). The designs show an _Update_ button—should that trigger the switch to the cart?
<img width="699" alt="screen shot 2017-06-27 at 12 14 47" src="https://user-images.githubusercontent.com/7767559/27585041-7547b3dc-5b32-11e7-9d12-d6f2a35c90b4.png">

tyxla commented:
> It seems slightly odd that changing the value of a dropdown whisks you off to another location (the cart). The designs show an Update button—should that trigger the switch to the cart?

It felt a little odd to me to have that Update button there in the first place, but now that you're saying it that way, it makes total sense. @rickybanister what do you think about this one?

seear commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,113 @@
+				args: {
+					timeSinceExpiry: purchase.expiryMoment.fromNow(),
+				},
+				context: 'timeSinceExpiry is of the form "[number] [time-period] ago" i.e. "3 days ago"',
```
This should likely be `comment`, not `context`. `context` is for when identical english phrases need different translations in some languages because of differing contexts.

tyxla replied:
Sharp-eye there! Fixed in https://github.com/Automattic/wp-calypso/pull/14856/commits/709c982f563778fa81b2f2520e08247b6e7f8bcd.

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,113 @@
+			} );
+		}
+
+		return translate( 'Billed yearly' );
```
Hi! I've found a possible matching string that has already been translated **22** times:
`translate( ', billed yearly' )` *ES Score: 20*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-plan-purchases-billing-period-switch)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

rickybanister commented:
So I think two things have changed since my original design was completed—

1. I didn't anticipate being taken to the cart from here
2. I didn't realize we were only allowing one-way changes from monthly to yearly

In lieu of those I think we need to change the UI. 

1. Let's completely forget about the 'Update' button in the section header. 
2. Let's have a confirmation dialog for when the plan owner is changed so a user doesn't lose access to the plan by accident
3. Let's switch from a dropdown to a primary button that says 'upgrade to yearly' (that's what the button says in the plans grid, right? let's make them match). A button will make sense if the user is taken to the cart.

Sorry for the late changes, but we are working with new information here.

tyxla commented:
Thanks so much for chiming in @rickybanister! I think what you suggested makes much more sense than what we had.

> Let's completely forget about the 'Update' button in the section header.

Sure. It won't be necessary anymore anyway.

> Let's have a confirmation dialog for when the plan owner is changed so a user doesn't lose access to the plan by accident

Sure, that would be nice, but it's outside of this PR's scope. This PR does not introduce a field for changing the plan ownership at all, so we don't need a confirmation dialog, right?

> Let's switch from a dropdown to a primary button that says 'upgrade to yearly' (that's what the button says in the plans grid, right? let's make them match). A button will make sense if the user is taken to the cart.

That makes much more sense indeed. Yes, we say "Upgrade to Yearly" in #14850, so I've used that for the button, here is an updated preview:

![](https://cldup.com/RkXeUDzQmc.png)

I could use another round of design and code feedback, @rickybanister @sirreal @seear.

rickybanister commented:
Now that I see it I think we could use some more context here. We should probably show the current billing period if it is Monthly with the button next to it. 

Here's a mockup:

![image](https://user-images.githubusercontent.com/437258/27978115-cb99d57e-633b-11e7-8a41-639b74b27148.png)

Compact, primary button, 12px margin between text and button, remove `vertical-align:baseline;` from `p .button` style declaration and it should look like my mockup.

tyxla commented:
Makes sense. Updated the PR, here is a preview of how it looks now:

![](https://cldup.com/5p-9XcUdaW.png)

@rickybanister let me know if you feel this needs further adjustment.

a8ci18n commented on client/me/purchases/manage-purchase/plan-details/billing-period.jsx:
```diff
@@ -0,0 +1,117 @@
+
+		return (
+			<FormSettingExplanation>
+				{ translate( 'Billed monthly' ) }
```
Hi! I've found a possible matching string that has already been translated **21** times:
`translate( 'per month, billed monthly' )` *ES Score: 14*
See 1 additional suggestion in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/jetpack-plan-purchases-billing-period-switch)

Help me improve these suggestions: react with :thumbsdown: if the suggestion doesn't make any sense, or with :thumbsup: if it's a particularly good one (even if not implemented).

sirreal commented on client/me/purchases/manage-purchase/plan-details/index.jsx:
```diff
@@ -67,10 +68,12 @@ class PurchasePlanDetails extends Component {
 		} );
 
 		return (
-			<div>
```
Alternatively, we can get something similar with something like this and no class or CSS needed:

```js
{ translate( 'Billed monthly' ) }
{ '  ' }
<Button
	onClick={ this.handleMonthlyToYearlyButtonClick }
	primary
	compact
>
```

tyxla replied:
No please. A dozen of little kittens cry every time we do it like this. 😿 

I agree it's sometimes easier to do it this way, but doing it with CSS gives us a lot more control over how it looks, and it allows us to target various device sizes and viewports separately. I'll keep the initial implementation.

sirreal replied:
👍  I'm good with the CSS implementation.

I do think we could have a conversation about the pros and cons of adding literal space between inline element 🙂

richardmuscat commented:
LGTM

rickybanister commented:
Yup, that looks perfect now @tyxla!

Sorry for my delayed response!

johnHackworth commented:
this looks good, let's merge it! :ship:
